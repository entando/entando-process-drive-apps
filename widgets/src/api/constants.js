export const DOMAINS = {
  PDA: process.env.REACT_APP_DOMAIN || '',
  APP_BUILDER: process.env.REACT_APP_APP_BUILDER_DOMAIN || '',
};

export const IS_MOCKED_API = process.env.REACT_APP_MOCKED_API === 'true';

export const MOCK_API_DELAY = 800;

export const LOCAL = process.env.REACT_APP_LOCAL === 'true';

export const METHODS = {
  GET: 'GET',
  POST: 'POST',
  PUT: 'PUT',
  DELETE: 'DELETE',
};

// Inject ur access_token if you are trying to reach the APP-BUILDER API locally
export const TOKENS = {
  APP_BUILDER: '0694821a5bdc2229aac820e8a773669a',
};
