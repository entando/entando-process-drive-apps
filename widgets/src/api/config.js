import { DOMAIN, IS_MOCKED_API, MOCK_API_DELAY } from 'api/constants';
import utils from 'utils';
import config from 'mocks/config';

const get = async widgetCode => {
  if (IS_MOCKED_API) {
    await utils.timeout(MOCK_API_DELAY);
    return config;
  }

  const url = `${DOMAIN}/configs/${widgetCode}`;
  const response = await fetch(url);

  return response.json();
};

export default {
  get,
};
