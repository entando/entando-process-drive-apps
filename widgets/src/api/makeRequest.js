import { IS_MOCKED_API, MOCK_API_DELAY, DOMAINS, METHODS } from 'api/constants';
import utils from 'utils';

const getParams = queryParams =>
  Object.keys(queryParams)
    .filter(k => queryParams[k] !== undefined && queryParams[k] !== '')
    .map(k => `${encodeURIComponent(k)}=${encodeURIComponent(queryParams[k])}`)
    .join('&');

export default async ({
  domain,
  uri,
  method,
  mockResponse = {},
  useAuthentication,
  body,
  headers = {},
  queryParams,
  forceMock,
}) => {
  if (IS_MOCKED_API || forceMock) {
    await utils.timeout(MOCK_API_DELAY);
    return mockResponse;
  }

  const url = `${domain || DOMAINS.PDA}${uri}${queryParams ? `?${getParams(queryParams)}` : ''}`;

  const requestHeaders = useAuthentication
    ? new Headers({
        'Content-Type': 'application/json',
        Authorization: `Bearer ${localStorage.getItem('token')}`,
        ...headers,
      })
    : new Headers(headers);

  const configs = {
    method: method || METHODS.GET,
    ...(body ? { body } : {}),
    headers: requestHeaders,
  };

  const response = await fetch(url, configs);
  if (!response.ok) {
    const payload = await response.json();
    throw new Error(
      payload.message
        ? payload.message
        : `Bad Response from server: ${response.status} ${response.statusText}`
    );
  }

  const responseHeaders = response.headers.get('Content-Type');

  if (responseHeaders && responseHeaders.includes('xml')) {
    return response.text();
  }
  if (responseHeaders && responseHeaders.includes('octet')) {
    return response;
  }
  return response.json();
};
