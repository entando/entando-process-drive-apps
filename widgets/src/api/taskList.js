import { IS_MOCKED_API, MOCK_API_DELAY, SERVICE } from 'api/constants';
import getTasks from 'mocks/taskList/api.mock';
import utils from 'utils';

const get = async (connection, page, pageSize, sortedColumn, sortOrder, filter) => {
  if (IS_MOCKED_API) {
    await utils.timeout(MOCK_API_DELAY);
    return getTasks(page, pageSize, sortedColumn, sortOrder, filter);
  }

  const url = `${SERVICE.URL}/connections/${connection}/tasks?sort=taskId`;
  const token = localStorage.getItem('token');
  const response = await fetch(url, {
    method: 'get',
    headers: new Headers({
      Authorization: `Bearer ${token}`,
      'Content-Type': 'application/json',
    }),
  });

  return response.json();
};

export default {
  get,
};
