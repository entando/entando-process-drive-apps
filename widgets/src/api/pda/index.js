import makeRequest from 'api/makeRequest';
import { SERVICE, METHODS } from 'api/constants';

import MOCK_CONNECTIONS from 'mocks/pda/connections';
import MOCK_PROCESSES from 'mocks/pda/processes';

export const getConnections = async () =>
  makeRequest({
    domain: SERVICE.URL,
    uri: '/connections',
    method: METHODS.GET,
    mockResponse: MOCK_CONNECTIONS,
    useAuthentication: false,
  });

export const getProcess = async connection =>
  makeRequest({
    domain: SERVICE.URL,
    uri: `/connections/${connection}/processes/definitions`,
    method: METHODS.GET,
    mockResponse: MOCK_PROCESSES,
    useAuthentication: false,
  });
