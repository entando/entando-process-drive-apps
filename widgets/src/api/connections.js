import { METHODS, DOMAINS } from 'api/constants';

import { CONNECTIONS } from 'mocks/taskList/configs';
import makeRequest from 'api/makeRequest';

// eslint-disable-next-line import/prefer-default-export
export const getConnections = async () =>
  makeRequest({
    domain: DOMAINS.PDA,
    uri: '/connections',
    method: METHODS.GET,
    mockResponse: CONNECTIONS,
    useAuthentication: false,
  });
