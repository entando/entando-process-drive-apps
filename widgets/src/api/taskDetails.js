import { SERVICE, METHODS } from 'api/constants';
import mockedTask from 'mocks/taskDetails/task';

import { CONNECTIONS, PROCESS, GROUPS, COLUMNS } from 'mocks/taskList/configs';
import makeRequest from 'api/makeRequest';

export const getTask = async (connection, taskId) => {
  return makeRequest({
    domain: SERVICE.URL,
    uri: `/connections/${connection}/tasks/${taskId}`,
    method: METHODS.GET,
    mockResponse: mockedTask,
    useAuthentication: true,
  });
};

// Configs

export const getConnections = async () =>
  makeRequest({
    domain: SERVICE.URL,
    uri: '/connections',
    method: METHODS.GET,
    mockResponse: CONNECTIONS,
    useAuthentication: false,
  });

export const getProcess = async connection =>
  makeRequest({
    // domain: PDA_DOMAIN,
    uri: `/connections/${connection}/processes/definitions`,
    method: METHODS.GET,
    mockResponse: PROCESS,
    useAuthentication: true,
  });

export const getGroups = async connection =>
  makeRequest({
    // domain: PDA_DOMAIN,
    uri: `/connections/${connection}/groups`,
    method: METHODS.GET,
    mockResponse: GROUPS,
    useAuthentication: true,
  });

export const getColumns = async connection =>
  makeRequest({
    // domain: PDA_DOMAIN,
    uri: `/connections/${connection}/tasks/columns`,
    method: METHODS.GET,
    mockResponse: COLUMNS,
    useAuthentication: true,
  });
