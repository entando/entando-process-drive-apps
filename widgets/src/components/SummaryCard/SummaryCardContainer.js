import React from 'react';
import { MuiThemeProvider as ThemeProvider } from '@material-ui/core/styles';
import PropTypes from 'prop-types';
import i18next from 'i18next';
import Paper from '@material-ui/core/Paper';
import Typography from '@material-ui/core/Typography';
import Divider from '@material-ui/core/Divider';
import Select from '@material-ui/core/Select';
import MenuItem from '@material-ui/core/MenuItem';
import withStyles from '@material-ui/core/styles/withStyles';
import { DOMAINS, LOCAL } from 'api/constants';
import theme from 'theme';
import { getPageWidget } from 'api/app-builder/pages';
import { getSummary } from 'api/pda/summary';
import CustomEventContext from 'components/SummaryCard/CustomEventContext';
import Skeleton from '@material-ui/lab/Skeleton';
import SummaryCardValues from 'components/SummaryCard/SummaryCardValues';

const styles = ({ palette }) => ({
  root: {
    background: '#FFF',
    borderTop: '4px solid #E7EAEC',
    color: '#676A6C',
  },
  header: {
    display: 'flex',
    padding: '20px 25px',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  headline: {
    fontWeight: 'bold',
  },
  periodSelectRoot: {
    border: `2px solid ${palette.primary.main}`,
    padding: '4px 24px 4px 14px',
    borderRadius: 4,
    fontSize: 10,
    fontWeight: 'bold',
    color: palette.primary.main,
  },
  periodSelectInputOutline: {
    '&$focused $notchedOutline': {
      border: 0,
    },
  },
  periodSelectIcon: {
    right: 3,
    color: palette.primary.main,
  },
  periodSelectItem: {
    fontSize: 12,
  },
});

class SummaryCard extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      config: null,
      loading: false,
      loadingValues: false,
      summary: null,
      period: 'monthly',
      trend: 'up',
    };

    this.fetchSummary = this.fetchSummary.bind(this);
    this.fetchWidgetConfigs = this.fetchWidgetConfigs.bind(this);
    this.handlePeriodChange = this.handlePeriodChange.bind(this);
  }

  componentDidMount() {
    const { serviceUrl } = this.props;
    if (!LOCAL) {
      // set the PDA domain to the URL passed via props
      DOMAINS.PDA = serviceUrl;
    }

    this.setState({ loading: true, loadingValues: true }, async () => {
      const { config: storedConfig } = this.state;
      const config = storedConfig || (await this.fetchWidgetConfigs());

      this.setState({ config }, () => this.fetchSummary());
    });
  }

  async fetchWidgetConfigs() {
    const { pageCode, frameId } = this.props;
    try {
      // config will be fetched from app-builder
      const widgetConfigs = await getPageWidget(pageCode, frameId);
      if (widgetConfigs.errors && widgetConfigs.errors.length) {
        throw widgetConfigs.errors[0];
      }

      const { config } = widgetConfigs.payload;

      return config;
    } catch (error) {
      this.handleError(error.message);
    }
    return null;
  }

  async fetchSummary() {
    const { config, period } = this.state;
    const { summaryId } = this.props;

    const connection = (config && config.knowledgeSource) || '';
    const containerId = (config && config.containerId) || '';
    const summaryContainerId = `${summaryId}@${containerId}`;

    try {
      const summary = await getSummary(connection, summaryContainerId, period);

      this.setState({
        loading: false,
        loadingValues: false,
        summary: (summary && summary.payload) || null,
      });
    } catch (error) {
      this.handleError(error.message);
    }
  }

  handleError(err) {
    const { onError } = this.props;
    onError(err);
  }

  handlePeriodChange(event) {
    this.setState({ period: event.target.value, loadingValues: true }, () => this.fetchSummary());
  }

  renderSkeletonHeader() {
    const { classes } = this.props;
    return (
      <div className={classes.header}>
        <Skeleton width="100px" height={27} variant="rect" />
        <Skeleton width="60px" height={27} variant="rect" />
      </div>
    );
  }

  renderHeader() {
    const { period, summary } = this.state;
    const { classes } = this.props;
    return (
      <div className={classes.header}>
        <Typography variant="subtitle2" component="h3" className={classes.headline}>
          {i18next.t(`card.labels.${summary.title}`)}
        </Typography>
        <Select
          value={period}
          variant="outlined"
          classes={{
            root: classes.periodSelectRoot,
            outlined: classes.periodSelectInputOutline,
            iconOutlined: classes.periodSelectIcon,
          }}
          onChange={this.handlePeriodChange}
        >
          {['monthly', 'annual', 'daily'].map(periodi => (
            <MenuItem key={periodi} value={periodi} className={classes.periodSelectItem}>
              {i18next.t(`card.${periodi}`)}
            </MenuItem>
          ))}
        </Select>
      </div>
    );
  }

  render() {
    const { loading, loadingValues, summary } = this.state;
    const { classes, width, display, onError } = this.props;

    return (
      <CustomEventContext.Provider value={{ onError }}>
        <ThemeProvider theme={theme}>
          <Paper square elevation={0} className={classes.root} style={{ width, display }}>
            {loading && this.renderSkeletonHeader()}
            {!loading && summary && this.renderHeader()}
            <Divider className={classes.divider} />
            <SummaryCardValues loading={loadingValues} values={summary} />
          </Paper>
        </ThemeProvider>
      </CustomEventContext.Provider>
    );
  }
}

SummaryCard.propTypes = {
  classes: PropTypes.shape({
    summaryCardWidgetBox: PropTypes.string,
  }).isRequired,
  width: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  display: PropTypes.string,
  summaryId: PropTypes.string.isRequired,
  onError: PropTypes.func,
  serviceUrl: PropTypes.string,
  pageCode: PropTypes.string,
  frameId: PropTypes.string,
};

SummaryCard.defaultProps = {
  onError: () => {},
  display: 'inline-block',
  width: 256,
  serviceUrl: '',
  pageCode: '',
  frameId: '',
  summaryId: 1,
};

export default withStyles(styles, { name: 'EntSummaryCard' })(SummaryCard);
