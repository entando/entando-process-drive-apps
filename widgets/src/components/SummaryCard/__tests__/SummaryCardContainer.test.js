import { render, wait } from '@testing-library/react';
import React from 'react';
import 'mocks/i18nMock';

import { DOMAINS } from 'api/constants';
import SummaryCardContainer from 'components/SummaryCard/SummaryCardContainer';
import mockSummary from 'mocks/summary/summary';
import WIDGET_CONFIGS from 'mocks/app-builder/pages';

describe('<SummaryCardContainer />', () => {
  it('renders snapshot correctly', async () => {
    const configUrl = `${DOMAINS.APP_BUILDER}/api/pages//widgets/`;
    const connection = 'kieStaging';
    const summaryUrl = `/connections/${connection}/summaries/summaryTypes/Card`;

    fetch.once(JSON.stringify(WIDGET_CONFIGS.SUMMARY_CARD)).once(JSON.stringify(mockSummary));

    const { container } = render(<SummaryCardContainer />);

    await wait(() => expect(container).toMatchSnapshot());

    expect(fetch.mock.calls.length).toBe(2);
    expect(fetch.mock.calls[0][0]).toEqual(configUrl);
    expect(fetch.mock.calls[1][0]).toEqual(summaryUrl);
  });

  it('renders snapshot correctly on error state', async () => {
    const { container } = render(<SummaryCardContainer />);

    await wait(() => expect(container).toMatchSnapshot());
  });
});
