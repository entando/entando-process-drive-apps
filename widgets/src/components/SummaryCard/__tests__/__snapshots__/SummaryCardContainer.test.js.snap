// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SummaryCardContainer /> renders snapshot correctly 1`] = `
<div>
  <div
    class="MuiPaper-root SummaryCard-root-1 MuiPaper-elevation0"
  >
    <div
      class="SummaryCard-header-2"
    >
      <span
        class="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-pulse"
        style="width: 100px; height: 27px;"
      />
      <span
        class="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-pulse"
        style="width: 60px; height: 27px;"
      />
    </div>
    <hr
      class="MuiDivider-root"
    />
    <div
      class="SummaryCardValues-root-42 upTrend"
    >
      <div>
        <span
          class="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-pulse"
          style="width: 110px; height: 37px;"
        />
        <span
          class="MuiSkeleton-root MuiSkeleton-rect SummaryCardValues-totalLabel-43 MuiSkeleton-pulse"
          style="width: 90px; height: 14px;"
        />
      </div>
      <div>
        <span
          class="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-pulse"
          style="width: 50px; height: 16px;"
        />
      </div>
    </div>
  </div>
  <div
    class="MuiSnackbar-root MuiSnackbar-anchorOriginBottomCenter"
  >
    <div
      class="MuiPaper-root MuiSnackbarContent-root ErrorNotification-error-49 MuiPaper-elevation6"
      direction="up"
      role="alert"
      style="opacity: 1; transform: scale(1, 1); transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    >
      <div
        class="MuiSnackbarContent-message"
      >
        <span
          class="ErrorNotification-message-46"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root ErrorNotification-icon-47 ErrorNotification-iconVariant-48"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-2h2v2zm0-4h-2V7h2v6z"
            />
          </svg>
          invalid json response body at undefined reason: Unexpected end of JSON input
        </span>
      </div>
      <div
        class="MuiSnackbarContent-action"
      >
        <button
          aria-label="close"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root ErrorNotification-icon-47"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`<SummaryCardContainer /> renders snapshot correctly on error state 1`] = `
<div>
  <div
    class="MuiPaper-root SummaryCard-root-99 MuiPaper-elevation0"
  >
    <div
      class="SummaryCard-header-100"
    >
      <span
        class="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-pulse"
        style="width: 100px; height: 27px;"
      />
      <span
        class="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-pulse"
        style="width: 60px; height: 27px;"
      />
    </div>
    <hr
      class="MuiDivider-root"
    />
    <div
      class="SummaryCardValues-root-140 upTrend"
    >
      <div>
        <span
          class="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-pulse"
          style="width: 110px; height: 37px;"
        />
        <span
          class="MuiSkeleton-root MuiSkeleton-rect SummaryCardValues-totalLabel-141 MuiSkeleton-pulse"
          style="width: 90px; height: 14px;"
        />
      </div>
      <div>
        <span
          class="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-pulse"
          style="width: 50px; height: 16px;"
        />
      </div>
    </div>
  </div>
  <div
    class="MuiSnackbar-root MuiSnackbar-anchorOriginBottomCenter"
  >
    <div
      class="MuiPaper-root MuiSnackbarContent-root ErrorNotification-error-147 MuiPaper-elevation6"
      direction="up"
      role="alert"
      style="opacity: 1; transform: scale(1, 1); transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    >
      <div
        class="MuiSnackbarContent-message"
      >
        <span
          class="ErrorNotification-message-144"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root ErrorNotification-icon-145 ErrorNotification-iconVariant-146"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-2h2v2zm0-4h-2V7h2v6z"
            />
          </svg>
          invalid json response body at undefined reason: Unexpected end of JSON input
        </span>
      </div>
      <div
        class="MuiSnackbarContent-action"
      >
        <button
          aria-label="close"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root ErrorNotification-icon-145"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
